lib/kafka_producer.js

```diff
@@ -111,99 +117,131 @@ KafkaProducer.prototype.connect = function connect(onConnect) {
+        } // TODO: else { self.logger.error() }
+        return;
+    }
+
     var messageBatch;
     var messageLength = message.length;
     var produceMessage = function produceMessage(msg) {
         return self.getProduceMessage(topic, msg, timeStamp, 'batch');
     };
 
-    if (self.topicToBatchQueue[topic]) {
+    if (!self.topicToBatchQueue[topic]) {
+        messageBatch = new MessageBatch(self.maxBatchSizeBytes);
+        self.topicToBatchQueue[topic] = messageBatch;
+    } else {
         messageBatch = self.topicToBatchQueue[topic];
-        if ((messageBatch.sizeBytes + messageLength + 4) > self.maxBatchSizeBytes) {
-            self.restClient.produce(produceMessage(messageBatch.getBatchedMessage()), function handleResponse(err, res) {
-                if (callback) {
-                    callback(err, res);
-                }
-            });
+    }
 
-            self.topicToBatchQueue[topic].resetBatchedMessage();
-        }
+    if ((messageBatch.sizeBytes + messageLength + 4) > self.maxBatchSizeBytes) {
+        self._produceBatch(messageBatch, topic, timeStamp);
+    }
 
-        // Do not add to buffer if message is larger than max buffer size, produce directly
-        if (messageLength > self.maxBatchSizeBytes) {
-            self.restClient.produce(produceMessage(message), function handleResponse(err, res) {
-                if (callback) {
-                    callback(err, res);
-                }
-            });
-        } else {
-            self.topicToBatchQueue[topic].addMessage(message, callback);
-        }
+    // Do not add to buffer if message is larger than max buffer size, produce directly
+    if (messageLength > self.maxBatchSizeBytes) {
+        self._produce(produceMessage(message), callback);
     } else {
-        messageBatch = new MessageBatch(self.maxBatchSizeBytes);
+        self.topicToBatchQueue[topic].addMessage(message, callback);
+    }
+};
 
-        // Do not add to buffer if message is larger than max buffer size, produce directly
-        if (messageLength > self.maxBatchSizeBytes) {
-            self.restClient.produce(produceMessage(message), function handleResponse(err, res) {
-                if (callback) {
-                    callback(err, res);
-                }
-            });
-        } else {
-            messageBatch.addMessage(message, callback);
+KafkaProducer.prototype._produceBatch =
+function _produceBatch(messageBatch, topic, timeStamp, callback) {
+    var self = this;
+
+    var msg = messageBatch.getBatchedMessage();
+    var pmsg = self.getProduceMessage(topic, msg, timeStamp, 'batch');
+
+    self._produce(pmsg, onProduced);
+
+    function onProduced(err, res) {
+        for (var i = 0; i < messageBatch.pendingCallbacks.length; i++) {
+            messageBatch.pendingCallbacks[i](err, res);
         }
 
-        self.topicToBatchQueue[topic] = messageBatch;
-    }
+        messageBatch.resetBatchedMessage();
```

Comment:
This reset needs to go after line 198 ```var msg = messageBatch.getBatchedMessage();```. If you wait until the callback is run, it will break because the offset isn't reset for new messages.